src/iop/toneequal.c

```diff
@@ -0,0 +1,1039 @@
+}
+
+
+#ifdef _OPENMP
+#pragma omp declare simd
+#endif
+static inline float RGB_light(const float pixel[4], const dt_iop_toneequalizer_method_t method)
+{
+  /** compute an estimation of the pixel lightness using several color models **/
+
+  if(method == DT_TONEEQ_MEAN)
+  {
+    /** Pixel intensity **/
+    return _RGB_mean(pixel);
+  }
+
+  else if(method == DT_TONEEQ_LIGHTNESS)
+  {
+    /** Pixel HSL lightness **/
+    return _RGB_lightness(pixel);
+  }
+
+  else if(method == DT_TONEEQ_VALUE)
+  {
+    /** Pixel HSV value **/
+    return _RGB_value(pixel);
+  }
+
+  else if(method == DT_TONEEQ_NORM_1)
+  {
+    /** Energy of the pixel **/
+    return _RGB_norm_1(pixel);
+  }
+
+  else if(method == DT_TONEEQ_NORM_2)
+  {
+    /** Radius of the pixel **/
+    return _RGB_norm_2(pixel);
+  }
+
+  else if(method == DT_TONEEQ_NORM_POWER)
+  {
+    /** Black magic : means nothing but looks good **/
+    return _RGB_norm_power(pixel);
+  }
+  return -1.0;
+}
+
+
+static inline void image_luminance(const float *const restrict in, float *const restrict out,
+                                   size_t width, size_t height, const dt_iop_toneequalizer_method_t method)
+{
+  /** Build the luminance map of an RGBA picture.
+   * The output is monochrome,  **/
+
+  const size_t ch = 4;
+
+#ifdef _OPENMP
+#pragma omp parallel for simd collapse(2) schedule(static)
+#endif
+  for(size_t i = 0; i < height; ++i)
+  {
+    for(size_t j = 0; j < width; ++j)
+    {
+      const size_t index = (i * width + j);
+      const float *pixel_in = __builtin_assume_aligned(in + index * ch, 16); // RGBa
+      float *pixel_out = __builtin_assume_aligned(out + index, 4); // single channel
+      *pixel_out = RGB_light(pixel_in, method);
+    }
+  }
+}
+
+
+#ifdef _OPENMP
+#pragma omp declare simd
+#endif
+static float compute_correction(const float factors[PIXEL_CHAN], const float exposure)
+{
+  // build the correction for the current pixel
+  // as the sum of the contribution of each luminance channel
+  float correction = 0.0f;
+  float weights[PIXEL_CHAN] __attribute__((aligned(64)));
+
+#ifdef _OPENMP
+#pragma omp simd
+#endif
+  for (int c = 0; c < PIXEL_CHAN; ++c) weights[c] = gauss(centers[c], exposure);
+
+#ifdef _OPENMP
+#pragma omp simd reduction(+:correction)
+#endif
+  for (int c = 0; c < PIXEL_CHAN; ++c) correction += weights[c] * factors[c];
+
+  return correction /= w_sum;
+}
+
+
+static inline void process_image(const float *const restrict in, float *const restrict out,
+                                  size_t width, size_t height,
+                                  const float factors[PIXEL_CHAN], const dt_iop_toneequalizer_method_t method,
+                                  float *const restrict luminance)
+{
+  const size_t ch = 4;
+
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static)
+#endif
+  for(size_t k = 0; k < (size_t)ch * width * height; k += 4 * ch)
```

Comment:
Should this be k += ch...

Reply:
I process 4 pixels at each loop step to vectorize the `fmaxf(log2f(...)...)` but, indeed, when the pixel number is not a multiple of 4, that could end out of bounds